runtime/strategies/create-handle-group.js

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {Strategy} from '../../strategizer/strategizer.js';
+import Recipe from '../recipe/recipe.js';
+import RecipeWalker from '../recipe/walker.js';
+import Handle from '../recipe/handle.js';
+
+export default class CreateHandleGroup extends Strategy {
+
+  async generate(inputParams) {
+    return Recipe.over(this.getResults(inputParams), new class extends RecipeWalker {
+      onRecipe(recipe) {
+        // Resolve constraints before assuming connections are free.
+        if (recipe.connectionConstraints.length > 0) return;
+
+        const freeConnections = recipe.handleConnections.filter(hc => !hc.handle && !hc.isOptional);
+        let maximalGroup = null;
+        for (let writer of freeConnections.filter(hc => hc.isOutput)) {
+          let compatibleConnections = [writer];
+          let involvedParticles = new Set([writer.particle]);
+
+          let foundSomeReader = false;
+          for (let reader of freeConnections.filter(hc => hc.isInput && hc !== writer)) {
+            if (Handle.effectiveType(null, [reader, ...compatibleConnections])
```

Comment:
I'm not sure it matters to begin to optimize this, but I think it should be possible to take a previous result of effectiveType and test it incrementally against the single new type.

Reply:
Done :)